File: _python/202.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        s = set()
        while n!=1:
            n=self.countN(n)
            if n not in s:
                s.add(n)
            else:
                return False
        return True

    @staticmethod
    def countN(n):
        res = 0
        while n != 0:
            tmp = (n % 10)**2
            res += tmp
            n = n / 10
        return res


if __name__ == '__main__':
    print Solution().isHappy(19)

File: _cpp/answer/202.cpp
//
// Created by wxy on 2018/2/15.
//
#include <set>

using namespace std;
namespace p202 {
    class Solution {
    public:
        bool isHappy(int n) {
            set<int> s;
            while (n != 1) {
                n = countN(n);
                if (s.find(n) == s.end())
                    s.insert(n);
                else
                    return false;
            }
            return true;
        }

        int countN(int n) {
            int res = 0, tmp;
            while (n != 0) {
                tmp = (n % 10);
                res += tmp * tmp;
                n = n / 10;
            }
            return res;
        }
    };
}